### ISSUE TITLE
Add event/message lister when db is closed.

### ISSUE THREAD

**Issue by The-CJ** [The-CJ/PhaazeDB#22]
Closed not from the WebUI,
currently with a high ore infinite save_interval it losses data when shut down.
Must be fixed

**Comment by The-CJ** [comment]
I mean things like closing the terminal, using systemctl or so
SYSINT, SYSTERM and so on

**Comment by The-CJ** [comment]
i have no idea how signal handler work... help :c

**Comment by HeapUnderfl0w** [comment]
https://docs.python.org/3.7/library/signal.html ?

**Comment by The-CJ** [comment]
Tryed
<pre>
	def start(self):
		self.Logger.info(f"Starting PhaazeDB v{self.version}")
		self.Logger.info(f"Running on port: {self.port}")
		if self.allowed_ips: self.Logger.info(f"Allowed IP's: {self.allowed_ips}")
		self.Logger.info(f"Action Logging: {self.action_logging}")

		signal.signal(signal.SIGTERM, self.signal_stop)
		signal.signal(signal.SIGINT, self.signal_stop)

		web.run_app(self.Server, port=self.port, print=False)

	def signal_stop(self):
		asyncio.ensure_future(self.stop())

</pre>
never got called

**Comment by The-CJ** [comment]
Also tryed:
<pre>
	def start(self):
		self.Logger.info(f"Starting PhaazeDB v{self.version}")
		self.Logger.info(f"Running on port: {self.port}")
		if self.allowed_ips: self.Logger.info(f"Allowed IP's: {self.allowed_ips}")
		self.Logger.info(f"Action Logging: {self.action_logging}")

		atexit.register(self.signal_stop)

		web.run_app(self.Server, port=self.port, print=False)

	def signal_stop(self):
		asyncio.ensure_future(self.stop())

</pre>
never callled

**Comment by HeapUnderfl0w** [comment]
signal works as expected.

Content of sigtest.py
```py
#!/usr/bin/env python

import signal
import time
import os

def call_on_sig(a, b):
    print("Sig")
    print("a = {0}".format(a))
    print("b = {0}".format(b))

signal.signal(signal.SIGTERM, call_on_sig)

print("We are " + str(os.getpid()))
while (True):
    time.sleep(1)
```
prints
```
We are 8871
Sig
a = 15
b = <frame at 0x7f76d25889f8, file 'sigtest.py', line 16, code <module>>
```
when killed with `kill -s SIGTERM 8871`

you have to handle the `KeyboardInterrupt` for CTRL+C handling i assume

**Comment by The-CJ** [comment]
Same as before
<pre>
	def start(self):
		self.Logger.info(f"Starting PhaazeDB v{self.version}")
		self.Logger.info(f"Running on port: {self.port}")
		if self.allowed_ips: self.Logger.info(f"Allowed IP's: {self.allowed_ips}")
		self.Logger.info(f"Action Logging: {self.action_logging}")

		signal.signal(signal.SIGTERM, self._stop)

		web.run_app(self.Server, port=self.port, print=False)

	def _stop(self, a, b):
		self.Logger.info(f"Sig a: {a}")
		self.Logger.info(f"Sig b: {b}")
		self.Logger.info(f"Sig: {a}")


</pre>
nothing, never called

**Comment by HeapUnderfl0w** [comment]
2 things: 
1) the function for `signal.signal` never takes a `self`. it is standalone (and takes 2 parameters: signum & stackframe)
2) it will not trigger on CTRL+C which means you may have to wrap the `web.run_app` call in a try-except:

```py
try:
    web.run_app(self.Server, port=self.port, print=False)
except KeyboardInterrupt:
    stop(None, None)
```

**Comment by The-CJ** [comment]
<pre>
	def start(self):
		self.Logger.info(f"Starting PhaazeDB v{self.version}")
		self.Logger.info(f"Running on port: {self.port}")
		if self.allowed_ips: self.Logger.info(f"Allowed IP's: {self.allowed_ips}")
		self.Logger.info(f"Action Logging: {self.action_logging}")

		signal.signal(signal.SIGTERM, signal_stop)

		web.run_app(self.Server, port=self.port, print=False)

	async def stop(self): . . .

def signal_stop(a, b):
	print(f"{a} {b}")
</pre>
nothing, even with "standalone" function.
self, is not visible for other function and just referenced the object after __init__

**Comment by HeapUnderfl0w** [comment]
are you killing it with `kill -s SIGTERM {pid}` ?

**Comment by The-CJ** [comment]
not only, has to work on all OS.
must also work with closing the terminal on windows

**Comment by HeapUnderfl0w** [comment]
the program will **never** get a signal when the Console gets closed via the X (that's just not in it for the shell)

additionally, as i said before: if you want to handle **user cancle** under Python, as far as i can gather you **have** to use `except KeyboardInterrupt`

**Comment by The-CJ** [comment]
KeyboardInterrupt never gets called, app runs in a different loop and never got this error

**Comment by HeapUnderfl0w** [comment]
then it seems (to me) that what you are attempting to do is impossible in the given architecture.
(unless you have any way of carrying the KeyboardInterrupt signal to the desired location from the place its recieved at)

**Comment by The-CJ** [comment]
mmm.. well 
signal.signal at least workes on systemctl stop
which sould be enough...

**Comment by HeapUnderfl0w** [comment]
yes, as `systemctl stop`, dependent on how the unit is configured, uses the `SIGTERM` signal to, well, signal the process it should exit (afaik it even uses a `SIGKILL` after a grace period of 120 seconds or so)
